### PR TITLE
doc: Add LwM2M engine state machine

### PR DESCRIPTION
**Pull request by jarlamsa**
The state machine depicts the states of the lwm2m engine
and the spawned RD client events that will be forwarded
to lwm2m client if it has registered a callback-function.
Most of the events are just informs for the client and the
RD client underneath works and recovers automatically from
most cases.

Signed-off-by: Jarno Lamsa <jarno.lamsa@nordicsemi.no>

### CODE REVIEW

**Comment by jarlamsa** [comment]
@peknis @SeppoTakalo please review.

File: doc/connectivity/networking/api/lwm2m.rst [hunk 1]
```diff
@@ -410,6 +410,93 @@ value of 1 is ok here).
 
 For a more detailed LwM2M client sample see: :ref:`lwm2m-client-sample`.
 
+LwM2M engine and application events
+***********************************
+
+The Zephyr LwM2M engine defines events that can be sent back to the application via callback
+functions. The engine state machine shows when the events are spawned. Events depicted in
+the diagram are listed in the table. The events are prefixed with `LWM2M_RD_CLIENT_EVENT_`.
```
Comment (peknis) [inline doc/connectivity/networking/api/lwm2m.rst:418]:
```suggestion
The Zephyr LwM2M engine defines events that can be sent back to the application through callback
functions.
The engine state machine shows when the events are spawned.
Events depicted in the diagram are listed in the table.
The events are prefixed with `LWM2M_RD_CLIENT_EVENT_`.
```

File: doc/connectivity/networking/api/lwm2m.rst [hunk 2]
```diff
@@ -410,6 +410,93 @@ value of 1 is ok here).
...
+     - Description
+     - Actions
+   * - 0
+     - NONE
+     - No event
+     - Do nothing
+   * - 1
+     - BOOTSTRAP_REG_FAILURE
+     - Bootstrap registration failed.
+       Happens if timeout or failure in bootstrap registration.
```
Comment (peknis) [inline doc/connectivity/networking/api/lwm2m.rst:440]:
```suggestion
       Occurs if there is a timeout or failure in bootstrap registration.
```

File: doc/connectivity/networking/api/lwm2m.rst [hunk 3]
```diff
@@ -410,6 +410,93 @@ value of 1 is ok here).
...
+     - Retry bootstrap
+   * - 2
+     - BOOTSTRAP_REG_COMPLETE
+     - Bootstrap registration complete.
+       Happens after successful bootstrap registration.
```
Comment (peknis) [inline doc/connectivity/networking/api/lwm2m.rst:445]:
```suggestion
       Occurs after successful bootstrap registration.
```

File: doc/connectivity/networking/api/lwm2m.rst [hunk 4]
```diff
@@ -410,6 +410,93 @@ value of 1 is ok here).
...
+     - No actions needed
+   * - 3
+     - BOOTSTRAP_TRANSFER_COMPLETE
+     - Bootstrap finish command received from the server.
```
Comment (peknis) [inline doc/connectivity/networking/api/lwm2m.rst:449]:
```suggestion
     - Bootstrap complete command received from the server.
```

File: doc/connectivity/networking/api/lwm2m.rst [hunk 5]
```diff
@@ -410,6 +410,93 @@ value of 1 is ok here).
...
+     - No actions needed, client proceeds to registration.
+   * - 4
+     - REGISTRATION_FAILURE
+     - Registration to LwM2M server failed.
+       Happens if timeout or failure in the registration.
```
Comment (peknis) [inline doc/connectivity/networking/api/lwm2m.rst:454]:
```suggestion
       Occurs if there is a timeout or failure in the registration.
```

File: doc/connectivity/networking/api/lwm2m.rst [hunk 6]
```diff
@@ -410,6 +410,93 @@ value of 1 is ok here).
...
+     - Retry registration
+   * - 5
+     - REGISTRATION_COMPLETE
+     - Registration to LwM2M server successful.
+       Happens after a successful registration reply from LwM2M server
```
Comment (peknis) [inline doc/connectivity/networking/api/lwm2m.rst:459]:
```suggestion
       Occurs after a successful registration reply from the LwM2M server
```

File: doc/connectivity/networking/api/lwm2m.rst [hunk 7]
```diff
@@ -410,6 +410,93 @@ value of 1 is ok here).
...
+       or when session resumption is used.
+     - No actions needed
+   * - 6
+     - REG_UPDATE_FAILURE
+     - Registration update failed.
+       Happens if a timeout happens during registration update.
```
Comment (peknis) [inline doc/connectivity/networking/api/lwm2m.rst:465]:
```suggestion
       Occurs if a there is a timeout during registration update.
```

File: doc/connectivity/networking/api/lwm2m.rst [hunk 8]
```diff
@@ -410,6 +410,93 @@ value of 1 is ok here).
...
+       NOTE: If registration update fails without a timeout,
+       a full registration is triggered automatically and
+       no registration update failure event is generated.
+     - No actions needed, client proceeds to re-registration automatically.
+   * - 7
+     - REG_UPDATE_COMPLETE
+     - Registration update completed.
+       Happens after successful registration update reply from LwM2M server.
```
Comment (peknis) [inline doc/connectivity/networking/api/lwm2m.rst:473]:
```suggestion
       Occurs after successful registration update reply from the LwM2M server.
```

File: doc/connectivity/networking/api/lwm2m.rst [hunk 9]
```diff
@@ -410,6 +410,93 @@ value of 1 is ok here).
...
+     - No actions needed
+   * - 8
+     - DEREGISTER_FAILURE
+     - Deregistration to LwM2M server failed.
+       Happens if timeout or failure in the deregistration.
+     - No actions needed, client proceeds to idle state automatically.
+   * - 9
+     - DISCONNECT
+     - Disconnected from LwM2M server.
+       Happens if timeout during discussing with server.
```
Comment (peknis) [inline doc/connectivity/networking/api/lwm2m.rst:483]:
```suggestion
       Occurs if there is a timeout during the dialogue with the server.
```

**Comment by peknis** [comment]
The image should be redrawn using Visio. Look for an example here: https://projecttools.nordicsemi.no/confluence/display/TECHDOC/State+diagram
And more figure style stuff here: https://projecttools.nordicsemi.no/confluence/display/TECHDOC/Figure+guide

**Review by peknis — CHANGES_REQUESTED** [review]

**Comment by SeppoTakalo** [comment]
> The image should be redrawn using Visio. Look for an example here...

This change is for public Zephyr project. I don't think internal Nordic guidelines apply. I would instead enforce the use of open source tools when submitting changes to open source community.

File: doc/connectivity/networking/api/lwm2m.rst (hunk 4 above)
Comment (SeppoTakalo) [reply doc/connectivity/networking/api/lwm2m.rst:449]: The operation is really named `Bootstrap-Finish`
https://www.openmobilealliance.org/release/LightweightM2M/V1_1_1-20190617-A/HTML-Version/OMA-TS-LightweightM2M_Core-V1_1_1-20190617-A.html#6-1-7-2-0-6172-Bootstrap-Finish-Operation

File: doc/connectivity/networking/api/lwm2m.rst (hunk 9 above)
Comment (SeppoTakalo) [reply doc/connectivity/networking/api/lwm2m.rst:483]: Maybe this should be `communication with the server`

File: doc/connectivity/networking/api/lwm2m.rst [hunk 10]
```diff
@@ -410,6 +410,93 @@ value of 1 is ok here).
...
+       Also triggered after deregistration has been done.
+     - No actions needed
```
Comment (SeppoTakalo) [inline doc/connectivity/networking/api/lwm2m.rst:485]: Should there be a mention that if connection is required, application should restart the client.

**Review by SeppoTakalo — DISMISSED** [review]

**Comment by jarlamsa** [comment]
Thank you for your reviews @SeppoTakalo @peknis, I think I have addressed your review comments and made the corrections in the document. Please re-review.

File: doc/connectivity/networking/api/lwm2m.rst [hunk 11]
```diff
@@ -410,6 +410,95 @@ value of 1 is ok here).
 
 For a more detailed LwM2M client sample see: :ref:`lwm2m-client-sample`.
 
+LwM2M engine and application events
+***********************************
+
+The Zephyr LwM2M engine defines events that can be sent back to the application via callback
+functions.
```
Comment (peknis) [inline doc/connectivity/networking/api/lwm2m.rst:417]:
```suggestion
The Zephyr LwM2M engine defines events that can be sent back to the application through callback
functions.
```

File: doc/connectivity/networking/api/lwm2m.rst [hunk 12]
```diff
@@ -410,6 +410,95 @@ value of 1 is ok here).
...
+The engine state machine shows when the events are spawned.
+Events depicted in the diagram are listed in the table.
+The events are prefixed with `LWM2M_RD_CLIENT_EVENT_`.
```
Comment (peknis) [inline doc/connectivity/networking/api/lwm2m.rst:420]:
```suggestion
The events are prefixed with ``LWM2M_RD_CLIENT_EVENT_``.
```

**Review by peknis — CHANGES_REQUESTED** [review]

**Review by peknis — APPROVED** [review]

**Review by SeppoTakalo — APPROVED** [review]

**Review by rlubos — APPROVED** [review]
LGTM, thanks!

**Pull request merged**